### PR TITLE
feat: update zod to allow custom viem chains in core

### PR DESCRIPTION
**Pull request by avasisht23**
More thorough refactor of: https://github.com/alchemyplatform/aa-sdk/pull/341

# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

### Focus of the PR:
This PR focuses on improving the ChainSchema in the `schema.ts` file and updating the AlchemyProvider to support chains that are not supported by Alchemy in the `base.ts` file.

### Detailed summary:
- Improved the ChainSchema in `schema.ts` by simplifying the custom validation logic.
- Updated the AlchemyProvider in `base.ts` to support chains that are not supported by Alchemy.
- Added a test case for validating the runtime when a chain is not supported by Alchemy in `light-account.test.ts`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/alchemy/src/provider/base.ts [hunk 1]
```diff
@@ -30,19 +29,14 @@ export class AlchemyProvider extends SmartAccountProvider<HttpTransport> {
     const config = AlchemyProviderConfigSchema.parse(config_);
 
     const { chain, entryPointAddress, opts, ...connectionConfig } = config;
-    const _chain =
-      typeof chain === "number" ? SupportedChains.get(chain) : chain;
```
Comment (denniswon) [inline packages/alchemy/src/provider/base.ts:34]: this is alchemy provider. don't we still need this as we have a notion of supported chains for alchemy?

**Review by denniswon — COMMENTED** [review]
did you mean to allow custom chain in the core provider rather than alchemy provider?

File: packages/alchemy/src/provider/base.ts (hunk 1 above)
Comment (avasisht23) [reply packages/alchemy/src/provider/base.ts:34]: correct. now these checks are in the `AlchemyChainSchema` which is called in `AlchemyProviderConfigSchema.parse` on line 30.

**Comment by avasisht23** [comment]
> did you mean to allow custom chain in the core provider rather than alchemy provider?

correct, I did enable custom chains in core providers, and still only supported chains with alchemy rpc urls in the viem object in alchemy providers.

File: packages/alchemy/src/provider/__tests__/light-account.test.ts [hunk 2]
```diff
@@ -132,6 +132,23 @@ describe("Alchemy Provider Tests", () => {
...
+      givenConnectedProvider({ owner, chain: avalanche });
+    }).toThrowErrorMatchingInlineSnapshot(`
+      "[
+        {
+          \\"code\\": \\"custom\\",
+          \\"fatal\\": true,
+          \\"path\\": [
+            \\"chain\\"
+          ],
+          \\"message\\": \\"Invalid input\\"
```
Comment (moldy530) [inline packages/alchemy/src/provider/__tests__/light-account.test.ts:147]: can we update the message to be a bit clearer here?

File: packages/alchemy/src/provider/__tests__/light-account.test.ts [hunk 3]
```diff
@@ -132,6 +132,23 @@ describe("Alchemy Provider Tests", () => {
...
   });
+
+  it("should correctly do runtime validation when chain is not supported by Alchemy", async () => {
+    expect(() => {
+      givenConnectedProvider({ owner, chain: avalanche });
+    }).toThrowErrorMatchingInlineSnapshot(`
+      "[
+        {
+          \\"code\\": \\"custom\\",
+          \\"message\\": \\"chain is not supported by Alchemy\\",
```
Comment (avarobinson) [inline packages/alchemy/src/provider/__tests__/light-account.test.ts:143]: Do we want to specify that this is based on viem supported chains?

**Review by avarobinson — APPROVED** [review]
One comment. Otherwise LGTM

File: packages/alchemy/src/provider/__tests__/light-account.test.ts (hunk 3 above)
Comment (avasisht23) [reply packages/alchemy/src/provider/__tests__/light-account.test.ts:143]: I think it's fine not to

**Review by moldy530 — DISMISSED** [review]

**Review by denniswon — APPROVED** [review]

**Pull request merged**